_posts/2021-05-21-dask-distributed-user-journey.md

```diff
@@ -0,0 +1,164 @@
+
+```python
+from dask.distributed import Client
+import dask.dataframe as dd
+
+client = Client()
+
+df = dd.read_csv('/data/.../2018-*-*.csv')
+df.groupby(df.account_id).balance.sum().compute()
+```
+
+## Breaking free from your machine
+
+Once you get used to task graphs and work scheduling you may begin thinking about how you can expand your computation beyond your computer.
+
+Our code doesn't really need to change much, we are already connecting a client and doing Dask work, all we need are more networked machines with the same user environments, data, etc.
+
+Personally I used to work in an organisation where every researcher was given a Linux desktop under their desk. These machines were on a LAN and had Active Directory and user home directories stored on a storage server. This meant you could sit down at any desk and log in and have a consistent experience. This also meant you could SSH to another machine on the network and your home directory would be there with all your files including your data and conda environments.
+
+This is a common setup in many organisations and it can be tempting to SSH onto the machines of folks who may not be fully utilising their machine and run your work there. And I'm sure you ask first right!
+
+Organisations may also have servers in racks designated for computational use and the setup will be similar. You can SSH onto them and home directories and data are available via network storage.
+
+With Dask Distributed you can start to expand your workload onto these machines using `SSHCluster`. All you need is your SSH keys set up so you can log into those machines without a password.
+
+```python
+from dask.distributed import Client, SSHCluster
+import dask.dataframe as dd
+
+cluster = SSHCluster(
+    [
+        "localhost",
+        "alices-desktop.lan",
+        "bobs-desktop.lan",
+        "team-server.lan",
+    ]
+)
+client = Client(cluster)
+
+df = dd.read_csv("/data/.../2018-*-*.csv")
+df.groupby(df.account_id).balance.sum().compute()
+```
+
+Now the same workload can run on all of the CPUs in our little ad-hoc cluster, using all the memory and pulling data from the same shared storage.
+
+## Moving to a compute platform
+
+Using (and abusing) hardware like desktops and shared servers will get you reasonably far, but probably to the dismay of your IT team.
+
+Organisations who have many users trying to perform large compute workloads will probably be thinking about or already have some kind of platform that is designated for running this work.
+
+The platforms your organisation has will be the result of many arbitrary technology choices. What programming languages does your company use? What deals did vendors offer at the time of procurement? What skills do the current IT staff have? What did your CTO have for breakfast the day they chose a vendor?
```

Comment:
Maybe "sometimes arbitrary" would be better here? Still cheeky, but leaves room to be less potentially insulting. 

Often a lot of the constraints that lead to particular decisions aren't known to users, so things can feel more arbitrary than they actually are.

Reply:
Yeah I know what you mean. I'm not sure "arbitrary" is necessarily the right word. But I couldn't think of a better one.

Folks make different decisions for various reasons. All reasons are generally valid. All choices are usually similar.

Decisions like Dell or HP for servers, AWS or GCP for cloud and Cray or IBM for HPC are usually not made because one technology is vastle superior to the other. They are made because there may be some small detail about your org which makes one choice slightly more compelling. These choices have a big impact on the details that engineers/scientists/etc have to care about, but very small impact in the grand scheme of things.

With Dask tooling we are attempting to homogenise those details that folks have to care about, so that the decision becomes "to Dask or not to Dask?" rather than "do my previous tech choices limit my ability to Dask?".

Reply:
Even after writing all that I'm still not sure if I can think of a better word than "arbitrary".